File: clients/ios/Sources/AutoUpload/Conditions.swift
import Foundation
import Network
import UIKit

/// Why a pass is not running right now — shown in the UI so "nothing happens" is explainable.
enum UploadBlock: Sendable, Equatable {
    case none
    case noNetwork
    case needsWiFi
    case notCharging
    case lowBattery
}

/// The network and power rules that gate a pass.
///
/// iOS has no roaming flag an app can read, so unlike Android there are three switches, not
/// four: `isExpensive` already covers cellular (and personal hotspots), which is what the
/// "Wi-Fi only" setting is really about.
final class Conditions: @unchecked Sendable {

    static let shared = Conditions()

    /// Below this an upload competes with the user's remaining battery.
    static let minBatteryPercent = 20

    private let monitor = NWPathMonitor()
    private let queue = DispatchQueue(label: "org.discodrive.conditions")
    private var observed: NWPath?

    private init() {
        monitor.pathUpdateHandler = { [weak self] p in self?.observed = p }
        monitor.start(queue: queue)
        DispatchQueue.main.async { UIDevice.current.isBatteryMonitoringEnabled = true }
    }

    /// The handler fires asynchronously, and `currentPath` is not usable either until the
    /// monitor has settled: a pass started from a freshly opened app read "no network" and
    /// did nothing — the very moment a user is most likely to press Upload now.
    private var path: NWPath { observed ?? monitor.currentPath }

    /// Waits briefly for the monitor to report a usable path. Callers that can await do so
    /// before checking, which turns a false "no network" into a short pause.
    func waitForPath(timeout: TimeInterval = 5) async {
        let deadline = Date().addingTimeInterval(timeout)
        while Date() < deadline {
            if path.status == .satisfied { return }
            try? await Task.sleep(nanoseconds: 200_000_000)
        }
    }

    func check(wifiOnly: Bool, chargingOnly: Bool, requireBattery: Bool) -> UploadBlock {
        let path = self.path
        guard path.status == .satisfied else { return .noNetwork }
        if wifiOnly && path.isExpensive { return .needsWiFi }

        let state = UIDevice.current.batteryState
        let plugged = state == .charging || state == .full
        if chargingOnly && !plugged { return .notCharging }
        if requireBattery && !plugged {
            let level = UIDevice.current.batteryLevel   // -1 when unknown
            if level >= 0 && Int(level * 100) < Self.minBatteryPercent { return .lowBattery }
        }
        return .none
    }
}

File: clients/android-discodrive/app/src/main/java/org/discodrive/android/autoupload/Conditions.kt
package org.discodrive.android.autoupload

import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.net.ConnectivityManager
import android.net.NetworkCapabilities
import android.os.BatteryManager
import org.discodrive.android.Prefs

/** Why a pass is not running right now — shown in the UI so "nothing happens" is explainable. */
enum class Block { NONE, NO_NETWORK, NEEDS_WIFI, ROAMING, NOT_CHARGING, LOW_BATTERY }

/**
 * The network and power rules that gate an upload pass. Global rather than per-rule: nobody
 * wants their camera folder on Wi-Fi only and their downloads folder on mobile data.
 */
object Conditions {
    /** Below this an upload competes with the user's remaining battery. */
    const val MIN_BATTERY_PERCENT = 20

    fun check(context: Context, prefs: Prefs): Block {
        val cm = context.getSystemService(ConnectivityManager::class.java)
        val caps = cm?.activeNetwork?.let { cm.getNetworkCapabilities(it) }
            ?: return Block.NO_NETWORK
        if (!caps.hasCapability(NetworkCapabilities.NET_CAPABILITY_INTERNET)) return Block.NO_NETWORK

        val unmetered = caps.hasCapability(NetworkCapabilities.NET_CAPABILITY_NOT_METERED)
        if (prefs.wifiOnly && !unmetered) return Block.NEEDS_WIFI
        if (prefs.pauseOnRoaming && !caps.hasCapability(NetworkCapabilities.NET_CAPABILITY_NOT_ROAMING)) {
            return Block.ROAMING
        }

        val status = context.registerReceiver(null, IntentFilter(Intent.ACTION_BATTERY_CHANGED))
        val plugged = (status?.getIntExtra(BatteryManager.EXTRA_PLUGGED, 0) ?: 0) != 0
        if (prefs.whileChargingOnly && !plugged) return Block.NOT_CHARGING
        if (prefs.requireBattery && !plugged) {
            val level = status?.getIntExtra(BatteryManager.EXTRA_LEVEL, -1) ?: -1
            val scale = status?.getIntExtra(BatteryManager.EXTRA_SCALE, -1) ?: -1
            if (level >= 0 && scale > 0 && level * 100 / scale < MIN_BATTERY_PERCENT) return Block.LOW_BATTERY
        }
        return Block.NONE
    }

    fun allow(context: Context, prefs: Prefs): Boolean = check(context, prefs) == Block.NONE
}
